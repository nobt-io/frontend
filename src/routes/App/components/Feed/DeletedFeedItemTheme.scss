.itemContentRoot {
  overflow: hidden;
  opacity: 0.3;
}

.itemText {
  overflow: hidden;
  text-overflow: ellipsis;
  text-decoration: line-through;
}
