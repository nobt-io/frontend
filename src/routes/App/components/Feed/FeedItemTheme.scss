.itemContentRoot {
  overflow: hidden;
}

.itemText {
  overflow: hidden;
  text-overflow: ellipsis;
}
