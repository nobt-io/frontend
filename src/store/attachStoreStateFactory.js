export default function(store) {
  return function attachStoreState(data, originalCallback = data => data) {
    data.extra.state = store.getState();
    return originalCallback(data);
  };
}
