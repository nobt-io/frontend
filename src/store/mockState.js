export class StateBuilder {

  App = {};
  router = {};

  /**
   * @param {AppStateBuilder} appStateBuilder
   * @returns {StateBuilder}
   */
  withApp = (appStateBuilder) => {
    this.App = appStateBuilder.build();
    return this;
  };

  build = () => {
    return {
      App: {...this.App},
      router: {...this.router}
    }
  }
}

export class AppStateBuilder {

  addBillForm = {};
  currentNobt = {};
  viewState = {};

  /**
   * @param {NobtStateBuilder} currentNobtStateBuilder
   * @return {AppStateBuilder}
   */
  withCurrentNobt = (currentNobtStateBuilder) => {
    this.currentNobt = currentNobtStateBuilder.build();
    return this;
  };

  build = () => {
    return {
      addBillForm: {...this.addBillForm},
      currentNobt: {...this.currentNobt},
      viewState: {...this.viewState}
    }
  }
}

export class NobtStateBuilder {

  data = {
    bills: [],
    payments: []
  };
  fetchNobtStatus = {};
  nobtFetchTimestamp = {};

  /**
   * @param id
   * @returns {NobtStateBuilder}
   */
  withId = (id) => {
    this.data.id = id;
    return this;
  };

  /**
   * @param name
   * @returns {NobtStateBuilder}
   */
  withName = (name) => {
    this.data.name = name;
    return this;
  };

  /**
   * @param currency
   * @returns {NobtStateBuilder}
   */
  withCurrency = (currency) => {
    this.data.currency = currency;
    return this;
  };

  /**
   * @param persons
   * @returns {NobtStateBuilder}
   */
  withParticipatingPersons = (...persons) => {
    this.data.persons = [ ...persons ];
    return this;
  };

  /**
   * @param status
   * @return {NobtStateBuilder}
   */
  withFetchNobtStatus = (status) => {
    this.fetchNobtStatus = status;
    return this;
  };

  /**
   * @param timestamp
   * @return {NobtStateBuilder}
   */
  withNobtFetchTimestamp = (timestamp) => {
    this.nobtFetchTimestamp = timestamp;
    return this;
  };

  /**
   * @param {BillBuilder} billBuilder
   * @returns {NobtStateBuilder}
   */
  withBill = (billBuilder) => {
    this.data.bills.push(billBuilder.build());
    return this;
  };

  /**
   * @param {PaymentBuilder} paymentBuilder
   * @returns {NobtStateBuilder}
   */
  withPayment = (paymentBuilder) => {
    this.data.payments.push(paymentBuilder.build());
    return this;
  };

  build = () => {
    return {
      data: {...this.data},
      fetchNobtStatus: this.fetchNobtStatus,
      nobtFetchTimestamp: this.nobtFetchTimestamp
    }
  }
}

class BillBuilder {

  shares = [];
  links = {};

  /**
   * @param {Number} id
   * @returns {BillBuilder}
   */
  withId(id) {
    this.id = id;
    return this;
  }

  /**
   * @returns {BillBuilder}
   */
  withCreationDate(date) {
    this.createdOn = date;
    return this;
  }

  /**
   * @param {String} href
   * @returns {BillBuilder}
   */
  withDeleteLink(href) {
    this.links.delete = {href};
    return this;
  }

  /**
   * @param {String} name
   * @returns {BillBuilder}
   */
  withName(name) {
    this.name = name;
    return this;
  }

  /**
   * @param {String} name
   * @returns {BillBuilder}
   */
  withDebtee(name) {
    this.debtee = name;
    return this;
  }

  /**
   * @param {String} debtor
   * @param {Number} amount
   * @returns {BillBuilder}
   */
  addShare(debtor, amount) {
    this.shares.push({debtor, amount});
    return this;
  }

  /**
   *
   * @returns {{id: *, createdOn: *}}
   */
  build() {
    return {
      id: this.id,
      createdOn: this.createdOn,
      name: this.name,
      debtee: this.debtee,
      shares: this.shares,
      _links: this.links
    }
  }
}

class PaymentBuilder {

  /**
   * @param {Number} id
   * @returns {PaymentBuilder}
   */
  withId(id) {
    this.id = id;
    return this;
  }

  /**
   * @returns {PaymentBuilder}
   */
  withCreationDate(date) {
    this.createdOn = date;
    return this;
  }

  /**
   * @param {String} name
   * @returns {PaymentBuilder}
   */
  withSender(name) {
    this.sender = name;
    return this;
  }

  /**
   * @param {String} name
   * @returns {PaymentBuilder}
   */
  withRecipient(name) {
    this.recipient = name;
    return this;
  }

  /**
   * @param {Number} amount
   * @returns {PaymentBuilder}
   */
  withAmount(amount) {
    this.amount = amount;
    return this;
  }

  /**
   *
   * @returns {{id: *, createdOn: *}}
   */
  build() {
    return {
      id: this.id,
      createdOn: this.createdOn,
      sender: this.sender,
      recipient: this.recipient,
      amount: this.amount
    }
  }
}

/**
 * @returns {NobtStateBuilder}
 */
export const aNobtState = () => new NobtStateBuilder()
  .withId("3XtC8MMj4bpY")
  .withName("Some nobt")
  .withCurrency("EUR")
  .withParticipatingPersons("Thomas", "David", "Philipp");

/**
 * @returns {AppStateBuilder}
 */
export const anAppState = () => new AppStateBuilder()
  .withCurrentNobt(aNobtState());

/**
 * @returns {StateBuilder}
 */
export const aStoreState = () => new StateBuilder().withApp(anAppState());

/**
 * @returns {BillBuilder}
 */
export const aBill = () => new BillBuilder();

/**
 * @returns {PaymentBuilder}
 */
export const aPayment = () => new PaymentBuilder();
