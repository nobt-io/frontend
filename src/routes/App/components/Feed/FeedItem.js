import { IconButton } from "react-toolbox/lib/button/index";
import { ListItem } from "react-toolbox/lib/list/index";
import FeedItemTheme from "./FeedItemTheme.scss"
import * as React from "react";

export default ({icon, caption, legend, onClick}) => (
  <ListItem
    key={caption + legend}
    leftIcon={icon}
    theme={FeedItemTheme}
    caption={caption}
    legend={legend}
    onClick={onClick && onClick}
    rightActions={[
      onClick && <IconButton icon="chevron_right" />
    ]}
    ripple={false}
  />
);
