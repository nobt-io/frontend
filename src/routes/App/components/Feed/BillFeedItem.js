import LocationBuilder from "../../modules/navigation/LocationBuilder";
import withNavigation from "../../../../components/hoc/withNavigation";
import * as React from "react";
import Amount from "../../../../components/Amount/Amount";
import FeedItem from "./FeedItem";

const BillFeedItem = ({feedItem, push}) => {

  const {id, debtee, subject, amount} = feedItem;

  return (
    <FeedItem
      icon="receipt"
      caption={`${debtee} paid '${subject}'`}
      legend={<Amount value={amount} />}
      onClick={() => LocationBuilder.fromWindow().push(id).apply(push)}
    />
  )
};

export default withNavigation(BillFeedItem);
