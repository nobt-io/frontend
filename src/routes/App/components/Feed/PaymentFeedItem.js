import * as React from "react";
import Amount from "../../../../components/Amount/Amount";
import FeedItem from "./FeedItem";

const PaymentFeedItem = ({feedItem}) => {

  const {sender, recipient, amount} = feedItem;

  return (
    <FeedItem
      icon="payment"
      caption={`${sender} paid ${recipient}`}
      legend={<Amount value={amount} />}
    />
  )

};

export default PaymentFeedItem;
