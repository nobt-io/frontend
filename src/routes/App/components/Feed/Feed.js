import * as React from "react";
import { ListSubHeader } from "react-toolbox/lib/list/index";
import HOList from "../../../../containers/HOList/HOList";
import { injectIntl } from "react-intl";
import { connect } from "react-redux";
import { getSortedFeedItems } from "../../modules/currentNobt/selectors";
import withNavigation from "../../../../components/hoc/withNavigation";
import BillFeedItem from "./BillFeedItem";
import PaymentFeedItem from "./PaymentFeedItem";

const FeedSection = ({name, feedItems, renderFeedItem}) => (
  <div>
    <ListSubHeader caption={name} />
    {
      feedItems.map(renderFeedItem)
    }
  </div>
);


class Feed extends React.Component {

  getFeedSections = () => {

    const now = this.props.intl.now();

    const formatFeedItemDate = feedItem => this.props.intl.formatRelative(feedItem.date, {now});

    const relativeDates = this.props.feedItems.map(formatFeedItemDate);
    const distinctRelativeDates = [ ...new Set(relativeDates) ];

    return distinctRelativeDates.map(date => {
      return {
        relativeDate: date,
        feedItems: this.props.feedItems.filter(feedItem => formatFeedItemDate(feedItem) === date)
      }
    })
  };

  static feedItemRenderer = {
    'bill': (feedItem) => <BillFeedItem feedItem={feedItem} />,
    'payment': (feedItem) => <PaymentFeedItem feedItem={feedItem} />
  };

  renderFeedItem = (feedItem) => {
    return Feed.feedItemRenderer[ feedItem.type ](feedItem);
  };

  render = () => (

    <HOList
      items={this.getFeedSections()}
      renderItem={section =>
        <FeedSection
          key={section.relativeDate}
          name={section.relativeDate}
          feedItems={section.feedItems}
          renderFeedItem={this.renderFeedItem}
        >
        </FeedSection>
      }
    />
  )
}

const FeedWithIntl = injectIntl(Feed);
const FeedWithNavigationAndIntl = withNavigation(FeedWithIntl);
const connectedComponentFactory = connect((state) => ({feedItems: getSortedFeedItems(state)}), {});

export default connectedComponentFactory(FeedWithNavigationAndIntl);
