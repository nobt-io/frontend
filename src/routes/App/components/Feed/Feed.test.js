import renderer from "react-test-renderer";
import { aBill, anAppState, aNobtState, aPayment } from "../../../../store/mockState";
import Feed from "./Feed";
import * as React from "react";
import { Provider } from "react-redux";
import { IntlProvider } from "react-intl";
import createMockStore from "../../../../store/mockStore";

describe('Feed', function () {

  it('should render feed correctly', function () {

    const SECOND = 1000;
    const MINUTE = SECOND * 60;
    const HOUR = MINUTE * 60;
    const DAY = HOUR * 24;

    const now = new Date(2017, 11, 9);

    let mockStore = createMockStore(
      stateBuilder => stateBuilder
        .withApp(
          anAppState()
            .withCurrentNobt(
              aNobtState()
                .withBill(
                  aBill()
                    .withCreationDate(now - DAY * 3)
                    .withName('Bier')
                    .withDebtee('Thomas')
                    .addShare('David', 10)
                    .addShare('Thomas', 10)
                    .addShare('Matthias', 10)
                )
                .withBill(
                  aBill()
                    .withCreationDate(now - DAY)
                    .withName('Tanken')
                    .withDebtee('David')
                    .addShare('David', 20)
                    .addShare('Thomas', 20)
                    .addShare('Matthias', 20)
                )
                .withBill(
                  aBill()
                    .withCreationDate(now - DAY)
                    .withName('Essen')
                    .withDebtee('Thomas')
                    .addShare('David', 15)
                    .addShare('Thomas', 15)
                    .addShare('Matthias', 15)
                )
                .withPayment(
                  aPayment()
                    .withCreationDate(now - DAY)
                    .withSender('David')
                    .withRecipient('Matthias')
                    .withAmount(15)
                )
            )
        )
    );

    let cmp = renderer.create(
      <IntlProvider initialNow={now}>
        <Provider store={mockStore}>
          <Feed />
        </Provider>
      </IntlProvider>
    );

    let tree = cmp.toJSON();

    expect(tree).toMatchSnapshot();

  });

});
