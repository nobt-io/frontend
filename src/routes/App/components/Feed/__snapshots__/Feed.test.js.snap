// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed should render feed correctly 1`] = `
<ul
  className=""
  data-react-toolbox="list"
>
  <div>
    <h5
      className=""
    >
      yesterday
    </h5>
    <li
      className="undefined "
      onClick={[Function]}
      onMouseDown={undefined}
      onTouchStart={undefined}
    >
      <span
        className=""
      >
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <span
              className="material-icons"
              data-react-toolbox="font-icon"
            >
              receipt
            </span>
          </span>
        </span>
        <span
          className=""
        >
          <span
            className="undefined"
            data-react-toolbox="list-item-text"
          >
            David paid 'Tanken'
          </span>
          <span
            className=""
            data-react-toolbox="list-item-text"
          >
            <span
              className=""
            >
              <span>
                €60.00
              </span>
            </span>
          </span>
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <button
              className=""
              data-react-toolbox="button"
              disabled={false}
              href={undefined}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchStart={[Function]}
              type="button"
            >
              <span
                className="material-icons"
                data-react-toolbox="font-icon"
              >
                chevron_right
              </span>
            </button>
          </span>
        </span>
      </span>
    </li>
    <li
      className="undefined "
      onClick={[Function]}
      onMouseDown={undefined}
      onTouchStart={undefined}
    >
      <span
        className=""
      >
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <span
              className="material-icons"
              data-react-toolbox="font-icon"
            >
              receipt
            </span>
          </span>
        </span>
        <span
          className=""
        >
          <span
            className="undefined"
            data-react-toolbox="list-item-text"
          >
            Thomas paid 'Essen'
          </span>
          <span
            className=""
            data-react-toolbox="list-item-text"
          >
            <span
              className=""
            >
              <span>
                €45.00
              </span>
            </span>
          </span>
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <button
              className=""
              data-react-toolbox="button"
              disabled={false}
              href={undefined}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchStart={[Function]}
              type="button"
            >
              <span
                className="material-icons"
                data-react-toolbox="font-icon"
              >
                chevron_right
              </span>
            </button>
          </span>
        </span>
      </span>
    </li>
    <li
      className="undefined "
      onClick={[Function]}
      onMouseDown={undefined}
      onTouchStart={undefined}
    >
      <span
        className=""
      >
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <span
              className="material-icons"
              data-react-toolbox="font-icon"
            >
              payment
            </span>
          </span>
        </span>
        <span
          className=""
        >
          <span
            className="undefined"
            data-react-toolbox="list-item-text"
          >
            David paid Matthias
          </span>
          <span
            className=""
            data-react-toolbox="list-item-text"
          >
            <span
              className=""
            >
              <span>
                €15.00
              </span>
            </span>
          </span>
        </span>
      </span>
    </li>
  </div>
  <div>
    <h5
      className=""
    >
      3 days ago
    </h5>
    <li
      className="undefined "
      onClick={[Function]}
      onMouseDown={undefined}
      onTouchStart={undefined}
    >
      <span
        className=""
      >
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <span
              className="material-icons"
              data-react-toolbox="font-icon"
            >
              receipt
            </span>
          </span>
        </span>
        <span
          className=""
        >
          <span
            className="undefined"
            data-react-toolbox="list-item-text"
          >
            Thomas paid 'Bier'
          </span>
          <span
            className=""
            data-react-toolbox="list-item-text"
          >
            <span
              className=""
            >
              <span>
                €30.00
              </span>
            </span>
          </span>
        </span>
        <span
          className={undefined}
        >
          <span
            className={undefined}
            onClick={undefined}
            onMouseDown={undefined}
          >
            <button
              className=""
              data-react-toolbox="button"
              disabled={false}
              href={undefined}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchStart={[Function]}
              type="button"
            >
              <span
                className="material-icons"
                data-react-toolbox="font-icon"
              >
                chevron_right
              </span>
            </button>
          </span>
        </span>
      </span>
    </li>
  </div>
</ul>
`;
